\documentclass[11pt]{article}
\usepackage[utf8]{inputenc}
\usepackage[T1]{fontenc}
\usepackage{minted}
\usepackage{multirow}
\usepackage{enumerate}
\usepackage{tikz}

%% for listings..
\newminted{perl}{linenos, bgcolor=mintedBg, fontsize=\footnotesize}
\newminted{r}{linenos, bgcolor=mintedBg, fontsize=\footnotesize}
\definecolor{mintedBg}{rgb}{0.95, 0.95, 0.95}
\definecolor{blockBg}{rgb}{0.6, 0.6, 0.95}

\begin{document}

\section{Bioinformatics in General}
\begin{enumerate}
\item What is bioinformatics and why is it becoming an essential topic for
  biologists?\\
  Note that the there is no single correct answer to this question, and that
  you will not be judged on what you argue, but on how reasonable and informed
  your argument is.
  
  In your answer you may wish to consider the following:
  \begin{itemize}
  \item your opinion of what bioinformatics is
  \item the technologies or skills that are used within the field
  \item the impact of technological process on biological experiments
  \end{itemize}
  (4 points)
\item Describe some examples of problems and fields addressed by
  Bioinformatics. Full marks require at least three examples; points
  will be given both for depth of explanation and for numbers of points.\\
  (4 points)
\item Why is there such an emphasis of the analysis of sequences within
  bioinformatics? (Any reasonable answer will be accepted).\\
  (2 points)
\item Describe and contrast how genes were identified before and after the
  availability of cheap DNA sequencing.\\
  (4 points)
\item What makes sequence data special?  [This can be phrased better: what quality
  of sequence data distinguishes it from others: qualitative allows identification]\\
  (2 points)
\item Describe an example of DNA sequence analysis that does not involve the
  alignment of sequences.\\
  (3 points)
\item What do we mean by, and when do we make use of sequence assembly?\\
  (3 points)
\item Give some examples of how we make use of sequence comparisons?\\
  (3 points)
\item What's the general relationship between sequence alignment and
  sequence comparison?\\
  (2 points)
\end{enumerate}

\section{Molecular Biology \& the Central Dogma}  
\begin{enumerate}
\item Describe how information stored in DNA sequence is transmitted and made use of in biological
  systems. What are the important properties of the different types of
  macro-molecules used in this process?
  \begin{itemize}
    \item What is DNA?
    \item How does it differ from RNA?
    \item Why does this matter?
    \item How do proteins differ from nucleic acids?
    \item What do we mean when we say that two nucleic acid sequences (RNA or
      DNA) are complementary?
    \item The majority of DNA molecules are double stranded, whereas the
      majority of RNA molecules are single stranded. Discuss the impact of
      strandedness (consider replication / structural properties).
    \item What is this concept generally called?
  \end{itemize}
  (10 points)
\item Describe the structure of genes in both prokaryotes and
  eukaryotes.Include as many features as you can.
  \begin{itemize}
  \item How do we usually define a gene? How does this differ
    from the historical definition of a gene.
  \item Discuss what should be included as part of a gene
    and how this makes it difficult to define the number of genes an organism
    has.
  \end{itemize}
  (10 points)
\item Give some exceptions to the Central Dogma.\\
(3 points)
\item What is the difference between RNA and DNA nucleotides and what
  consequence does it have?\\
(2 points)
\item How do the monomers of nucleic acids differ from those of proteins? How
  does this relate to their function?\\
(4 points)
\item What are exons and introns?\\
(3 points)
\item What is a codon? How long is it and why?\\
(4 points)
\item What do we mean when we talk about the degeneracy of the genetic code?\\
(2 points)
\item Given a genetic code:\\
  \begin{minipage}{0.6\textwidth}
  {\tiny
    \input{../tex_resources/genetic_code.tex}
  }
  \end{minipage}

  Give all the amino acide sequences that can be encoded by:\\
  \verb|ATCAGATAGATATTACCG|\\
  (4 points)
\item What is an open reading frame (ORF)? What good are they to
  bioinformaticians?\\
  (2 points)
\item What kind of mutations can arise in DNA sequences? What effects
  do the different types of mutations have on genes?\\
  (4 points)
\end{enumerate}

\section{Biological databases}
\begin{enumerate}
\item Describe the typical structure of a web-facing databases. Comment on the
  different types of users and on how they use the database differently.\\
  (3 points)
\item Describe examples of the types of data that is held in different
  databases. Comment on how the type of data affects how the data is stored
  and accessed.\\
  (3 points)
\item Where can you find DNA sequences? (If you have trouble remembering
  specific names of databases, describe in some detail how you would go about finding a
  location / database that gives you DNA sequences).\\
  (3 points)
\item How would you find biological information about a given DNA sequence?
  Include details about the type of information that can be encoded by a DNA
  sequence and how this relates to your choice of database.
  (remember the structure of genes).\\
  (3 points)
\item You search the [name witheld] database for a sequence representing a
  specific gene. You get several hundred sequences back. What are some of the reasons
  for the multitude of sequences? What can you do to obtain only relevant
  sequences? Finally, what questions might you be able to address with the
  full set of sequences you obtained?\\
  (3 points)
\item What are the differences between the Genbank, EMBL and DDBJ databases?\\
  (2 points)
\item Describe the fasta sequence format.\\
  (2 points)
\item Apart from sequence data what does the fastq format contain?\\
  (1 point)
\item What are genome databases and why are they useful?\\
  (4 points)
\end{enumerate}

\section{Pairwise alignment}
\begin{enumerate}
\item For the two sequences: \verb|5' TAATTAGC 3'| and \verb|5' CATTA 3'| :
  \begin{enumerate}
    \item Provide three reasonable alignments.
    \item Describe how you can score the alignments using sets of penalties
      (remember that you can have negative penalties).
    \item Score the alignments using at least two different sets of
      penalties. Comment on how the choice of penalties relates to the
      biological question being asked.
    \item Which of your alignments do you think is most reasonable and why? If you
      think all are equally reasonable then explain your opinion.
  \end{enumerate}
  (10 points)
\item Why do we often use a different penalty for gap insertion and gap
  extension? Provide an example alignment with your question.\\
  (4 points)
\item Describe the dotplot? What can we use it for?\\
  (4 points)
\item What do we mean by an optimal alignment?\\
  (2 points)
\item What is the difference between a global and local alignment?\\
  (3 points)
\item What is this?
  \begin{figure}[H]
    \begin{tikzpicture}[scale=0.6]
      \input{nm1_aa.tikz}
    \end{tikzpicture}
  \end{figure}
  \begin{enumerate}
  \item Fill in the next row with both numbers and arrows.
  \item What can you obtain from this table after it is completely filled in?
  \end{enumerate}
  (4 points)
\item Obtain all optimal alignments from the following table:
  \begin{figure}[H]
    \begin{tikzpicture}[scale=0.6]
      \input{nm2_aa.tikz}
    \end{tikzpicture}
  \end{figure}
  \begin{enumerate}
  \item Draw the process directly on the table.
  \item What kind of alignment(s) did you obtain?
  \end{enumerate}
  (4 points)
\item How would you modify the Needleman-Wunsch (global alignment) to
  obtain a local alignment (Smith-Waterman)?\\
  (3 points)
\item Fill in the first two rows and columns of the following
  alignment table for a local alignment (Smith-Waterman).
  \begin{figure}[H]
    \begin{tikzpicture}[scale=0.6]
      \input{sm1_aa.tikz}
    \end{tikzpicture}
  \end{figure}
  (3 points)
\item What is an amino acid substitution matrix and how would you use one?\\
  (3 points)
\item Describe at least two ways in which you can derive an amino acid
  substitution matrix?\\
  (4 points)
\item On what basis could you define a nucleotide substitution matrix for
  aligning DNA sequences?\\
  (2 points)
\item The following is an example of an amino acid substitution table.
  \begin{figure}[H]
    \includegraphics[width=0.7\textwidth]{images/dayhoff_256}
  \end{figure}
  \begin{enumerate}
  \item The table has high and low values. What do these mean?
  \item What do the numbers along the diagonal indicate? Why do they
    vary?
  \end{enumerate}
  (4 points)
\end{enumerate}
  
\section{Multiple sequence alignment}
\begin{enumerate}
\item What are homologues, orthologues and paralogues?\\
  (4 points)
\item Describe some uses of multiple sequence alignment.\\
  (3 points)
\item What are heuristic methods? Why do we use them?\\
  (3 points)
\item Describe the three steps of the Clustal method?\\
  (4 points)
\item How does Clustal method make use of a guide tree?\\
  (3 points)
\item Describe a simple method to obtain a guide tree based on pairwise
  distances.\\
  (3 points)
  
\end{enumerate}

\section{Perl}
\begin{enumerate}
\item What is Perl, and what is it useful for?\\
  (3 points)
\item What are the three ways in which you can store values in Perl? Give
  examples as to how you can assign values to such variables?\\
  (4 points)
\item What does the following code do?

  \begin{perlcode}
  #!/usr/bin/perl -w
    
  $a = $ARGV[0];
  $b = $ARGV[1];

  print "$a + $b = ", $a + $b, "\n";
    
  \end{perlcode}
  Explain what kind of errors you may get running the above code.\\
  (4 points)
\item What does the following code do?

  \begin{perlcode}
  #!/usr/bin/perl -w

  @a = (0,1);
  for($i = 2; $i < 100; $i++){
    $a[$i] = $a[$i-2] + $a[$i-1];
  }
  
  for $v(@a){
    print "$v\n";
  }
\end{perlcode}
(4 points)
\item The Fibonacci sequence is defined as a sequence of numbers starting
  with 1,1 or 0,1 and where every subsequent number is the sum of the two
  previous numbers. That is: $F_i = F_{i-2} + F_{i-1}$. Write a small piece
  of Perl that calculates and prints the first 100 values of the Fibonacci
  sequence. (You will be minimally penalised for syntax errors, the logic is
  more important).\\
  (5 points)
\item How would you represent a table of values in Perl?\\
  (2 points)
\item Why is the following problematic:

  \begin{perlcode}
  $v1 = "one";
  $v2 = 2;
  $v3 = $v1 + $v2;
  \end{perlcode}
  (2 points)
\item What is the difference between:

  \begin{perlcode}
  if($a eq $b)
  \end{perlcode}
  and

  \begin{perlcode}
  if($a == $b)
  \end{perlcode}
  When would you use one rather than the other?\\
  (2 points)
\item What will the following code print:

  \begin{perlcode}
  $a = 10;
  $b = 8;
  if($b = $a){
    print "$a is equal to $b\n";
  }else{
    print "$a is not equal to $b\n";
  }
  \end{perlcode}
  Read carefullly; there is a small trick here.\\
  (2 points)
\end{enumerate}

\section{Searching databases by sequence}
\begin{enumerate}
\item Describe the relationship between pairwise sequence alignment and
  searching a database of sequences for homologous sequences.\\
  (3 points)
\item Describe three examples of situations where you may wish to search a
  sequence database for homologous or overlapping sequences.\\
  (4 points)
\item When would you use Blat as opposed to Blast?\\
  (2 points)
\item Write down a lookup table (index) for the following sequence:
\begin{verbatim}
TPATGGANWT
\end{verbatim}
 (3 points)
\item Why do we not normally use a dynamic programming method to identify
  homologous sequences.\\
  (1 point)
\item You have searched a database for homologous sequences and have obtained
  the following results:
\begin{figure}[H]
  \input{database_alignments.tikz}
\end{figure}
{\small(db 1, db 2, etc..., refer to individual sequences within the database, the
thickness of the diagonal lines refer to local alignment scores)}\\
What can you infer from these alignments? How does the choice of database
affect the inferences that you can make?\\
(4 points)
\item Give three different types of nucleic acid sequences (from eukaryotes)
  depending on their relationship to genes. What are these types and how does
  it affect your choice of database to use when looking for similar sequences?\\
(3 points)
\item How would you identify an alignment between sequence 1 (\verb|FLWRTWS|) 
  and 2 (\verb|SWKTWT|) from the following table:
\begin{figure}[H]
  \input{fasta_table.tex}
\end{figure}
{\small Hints: It may help you to plot column \texttt{S1} vs
  \texttt{S2}.
}\\
(5 points)
\item There are 5 basic types of blast. Explain what these variants are and
  why they exist. You do not need to give the correct names, though you do
  need to explain the differences in what the programs do.\\
  (4 points)
\item The following shows several parts of a genbank formatted file. Lines
  containing only \texttt{...} indicate parts that have been removed from the
  file. What can you tell about the sequence given in the file?
  {\tiny \input{sequence_parts.gb}}
  (4 points)
\item The following image shows the graphical output from an NCBI blast
  search. What can you tell from the picture?
  \begin{figure}[H]
  \includegraphics[width=0.6\textwidth]{images/blast_result_image}
  \end{figure}
  (3 points)
\item The following is a screenshot of the output from an NCBI blast
  search. What do the different columns indicate (consider the total scores
  and max scores as a single score for the sake of simplicity). Which of these
  scores is usually the most important and why?
  \begin{figure}[H]
    \includegraphics[width=\textwidth]{images/blast_result_list_top}
  \end{figure}
  (3 points)
\end{enumerate}

\section{Numbers, big data and R}
\begin{enumerate}
\item Describe the three most commonly used average values.\\
  (3 points)
\item What is meant by 'the distribution of values' and how can you obtain
  this in \texttt{R}?\\
  (3 points)
\item What is meant by, a) a normal distribution and b) a log normal
  distribution?\\
  (3 points)
\item The following figure shows the distributions of two sets of values \texttt{A}
  and \texttt{B}.
  \begin{figure}[H]
    \includegraphics[width=0.8\textwidth]{R/log_norm.pdf}
  \end{figure}
  What type of distribution do you think the values of \texttt{A} follow? What are your
  reasons for thinking this? Estimate the values for the mean, median and mode
  averages for \texttt{B}.\\
  Suggest a distribution for \texttt{B}. Would it be reasonable to calculate
  average values for \texttt{B}? How would you transform the values of
  \texttt{B} before additional analyses?\\
  (4 points)
\item The following plot shows the distribution of a sampling statistic ($X$)
  derived from randomised values that correspond to what would be observed
  under the null hypothesis. 
  \begin{figure}[H]
    \includegraphics[width=0.5\textwidth]{R/hyper}
  \end{figure}
  How would you estimate the probability of
  observing a value of more 10?\\
  (hint: you don't need to plot the distribution, but it is shown as an aid to
  your thinking)\\
  (4 points)
\item Describe in words what the \emph{t-statistic} calculates.\\
  (2 points)
\item The t-statistic is often used to estimate the statistical significance
  of the difference of two means. It relies on the user being able to estimate
  the population variance of experimental measurements from a limited number
  of experimental samples.\\
  Consider the following scenario: you are measuring
  some variable $v$ that is influenced by your experimental conditions. Each
  measurement can be considered a random sample from the population distributions of
  $v$. The following plots show the population distributions of $v$ under the
  experimental conditions $A$ and $B$. The left and right panels show
  different types of distributions for $v$. 
  \begin{figure}[H]
    \includegraphics[width=0.8\textwidth]{R/binorm}
  \end{figure}
  Discuss the use of the t-test for
  data arising from such distributions. What should you do to avoid misleading
  statistics?\\
  (4 points)
\item What does the normal distribution indicate?\\
  (2 points)

\item What is R? What is it good and not so good for?\\
  (2 points)
\item Describe the main data types supported by R?\\
  (4 points)
\item What are the main data structures used in R?\\
  (4 points)

\item In the following R-code;

  \begin{rcode}
  ## exp.data is a matrix of expression values that 
  ## has been defined earlier in the analysis
  ## where each column represents a sample 
  ## and each row a gene transcript
  
  h.all <- hist( log(exp.data) )
  h.sample <- apply( log(exp.data), 2, function(x){ hist(x,
    breaks=h.all$breaks) })
  
  h.sample.c <- sapply( h.sample, function(x){ x$counts })
  \end{rcode}
  Describe what is obtained on each line of functional code (6, 7, 10).\\
  (4 points)
\item What does the following R-code do?

  \begin{rcode}
  a <- vector(mode='numeric', length=1000)
  b <- vector(mode='numeric', length=length(a))
  for(i in 1:length(a)){
    a[i] <- rnorm(1, 10, 2)
    b[i] <- a[i] * rnorm(1, 2, 0.25) + rnorm(1, 3, 1)
  }
  plot(a, b, pch=1, col='red', xlab='a', ylab='b')    
  \end{rcode}
  Describe what each line does, one by one.\\
  The above code is bad R-code that will run very slowly for larger data
  sets. How would you rewrite it to make it run much, much faster?\\
  Hint: \texttt{rnorm(n, m, sd)} is a function that provides 
  \texttt{n} random samples from a normal distribution with a mean
  of \texttt{m} and standard deviation of \texttt{sd}.\\
  (4 points)
\item The following code creates some matrices:
 
  \begin{rcode}
  a <- matrix(1:12, nrow=4)
  b <- a[1:2, ]
  c <- cbind(b, c(10,11))
  d <- a[ rowSums(a) > 18, ]
  \end{rcode}
  Draw the resulting data structures (a, b, c and d).\\
  (3 points)
\item What's the difference between a vector and a list?\\
  (2 points)
\item What's the most important R function?\\
  (1 point)
\end{enumerate}

\end{document}
